snapcraft/internal/project_loader/_config.py

```diff
@@ -218,6 +218,18 @@ def __init__(self, project: project.Project) -> None:
         self.build_tools = grammar_processor.get_build_packages()
         self.build_tools |= set(project.additional_build_packages)
 
+        # Always add the base for building for non os and base snaps
+        if project.info.base is not None and project.info.type not in ("base", "os"):
+            self.build_snaps.add(project.info.base)
+        elif project.info.type not in ("base", "os"):
+            # This exception is here to help with porting issues with bases. In normal
+            # executions, when no base is set, the legacy snapcraft will be executed.
+            raise RuntimeError(
+                "A base is required for the scenario of snap type {!r}".format(
```

Comment:
"A base is required for {!r} snaps." conveys the same information in less words.